source/Missions/Services/Agents/Patches/AgentDamagePatch.cs

```diff
@@ -16,25 +24,40 @@ public class AgentDamagePatch
     {
         static bool Prefix(Agent attacker, Agent victim, Blow b, ref AttackCollisionData collisionData)
         {
-            // first, check if the attacker exists in the agent to ID groud, if not, no networking is needed (not a network agent)
-            if (NetworkAgentRegistry.Instance.TryGetAgentId(attacker, out Guid attackerId) == false) return true;
-
-            // next, check if the attacker is one of ours, if not, no networking is needed (not our agent dealing damage)
-            if (NetworkAgentRegistry.Instance.IsControlled(attackerId) == false) return true;
-
-            // If there is package factory cannot be resolved, do default behavior
-            if (ContainerProvider.TryResolve<IBinaryPackageFactory>(out var packageFactory) == false) return true;
-
-            // get the victim GUI
-            NetworkAgentRegistry.Instance.AgentToId.TryGetValue(victim, out Guid victimId);
-
             // construct a agent damage data
-            AgentDamageData agentDamageData = new AgentDamageData(attackerId, victimId, collisionData, b);
+            AgentDamage agentDamageData = new AgentDamage(attacker, victim, b, collisionData);
 
             // publish the event
-            NetworkMessageBroker.Instance.PublishNetworkEvent(agentDamageData);
+            MessageBroker.Instance.Publish(attacker, agentDamageData);
+
+            return true;
+        }
+    }
+    [HarmonyPatch(typeof(Mission), "OnAgentHit")]
+    internal static class OnAgentHitPatch
+    {
+        private static bool Prefix(ref Agent affectorAgent)
+        {
+            // Only allow damage from controlled agents
+            if (!NetworkAgentRegistry.Instance.IsControlled(affectorAgent)) return false;
 
             return true;
         }
     }
+    [HarmonyPatch(typeof(Agent), nameof(Agent.ChangeWeaponHitPoints))]
+    public class ShieldDamagePatch
+    {
+        static void Postfix(Agent __instance, EquipmentIndex slotIndex, short hitPoints)
+        {
+            if (NetworkAgentRegistry.Instance.TryGetAgentId(__instance, out Guid agentId) == false) return;
+
+            if (NetworkAgentRegistry.Instance.IsControlled(agentId) == false) return;
+
+            if (hitPoints <= 0) 
+            {
+                ShieldBreak shieldDamage = new ShieldBreak(__instance, slotIndex);
+                NetworkMessageBroker.Instance.Publish(__instance, shieldDamage);
+            }
+        }
```

Comment:
Rename `ShieldBreakHandler` to `ShieldDamageHandler`, move this logic there, and subscribe to this event
```suggestion
            ShieldDamaged shieldDamaged = new ShieldDamaged(__instance, slotIndex);
            NetworkMessageBroker.Instance.Publish(__instance, shieldDamaged);
        }
```